pkg/doc/jose/jws.go

```diff
@@ -0,0 +1,380 @@
+	return s(joseHeaders, payload, signingInput, signature)
+}
+
+// CompositeAlgSignatureVerifier defines composite signature verifier based on the algorithm
+// taken from JOSE header alg.
+type CompositeAlgSignatureVerifier struct {
+	verifierByAlg map[string]SignatureVerifier
+}
+
+// NewCompositeAlgSignatureVerifier creates a new CompositeAlgSignatureVerifier
+func NewCompositeAlgSignatureVerifier() *CompositeAlgSignatureVerifier {
+	return &CompositeAlgSignatureVerifier{
+		verifierByAlg: make(map[string]SignatureVerifier),
+	}
+}
+
+// AddVerifier adds a new verifier of algorithm.
+// If algorithm verifier is already present, it's get rewritten.
+func (v *CompositeAlgSignatureVerifier) AddVerifier(
+	alg string, verifier SignatureVerifier) *CompositeAlgSignatureVerifier {
+	v.verifierByAlg[alg] = verifier
+	return v
+}
+
+// Verify verifiers JWS signature.
+func (v *CompositeAlgSignatureVerifier) Verify(joseHeaders Headers, payload, signingInput, signature []byte) error {
+	alg, ok := joseHeaders.Algorithm()
+	if !ok {
+		return errors.New("'alg' JOSE header is not present")
+	}
+
+	verifier, ok := v.verifierByAlg[alg]
+	if !ok {
+		return fmt.Errorf("no verifier found for %s algorithm", alg)
+	}
+
+	return verifier.Verify(joseHeaders, payload, signingInput, signature)
+}
+
+// Signer defines JWS Signer interface. It makes signing of data and provides custom JWS headers relevant to the signer.
+type Signer interface {
+	// Sign signs.
+	Sign(data []byte) ([]byte, error)
+
+	// Headers provides JWS headers. "alg" header must be provided (see https://tools.ietf.org/html/rfc7515#section-4.1)
+	Headers() Headers
+}
+
+// NewJWS creates JSON Web Signature.
+func NewJWS(protectedHeaders, unprotectedHeaders Headers, payload []byte) *JSONWebSignature {
+	return &JSONWebSignature{
+		ProtectedHeaders:   protectedHeaders,
+		UnprotectedHeaders: unprotectedHeaders,
+		Payload:            payload,
+	}
+}
+
+// SerializeCompact makes JWS Compact Serialization (https://tools.ietf.org/html/rfc7515#section-7.1)
+func (s *JSONWebSignature) SerializeCompact(signer Signer, detached bool) (string, error) {
+	err := s.sign(signer)
+	if err != nil {
+		return "", fmt.Errorf("calculate compact JWS signature: %w", err)
+	}
+
+	joseHeaders := mergeHeaders(s.ProtectedHeaders, signer.Headers())
+
+	byteHeaders, err := json.Marshal(joseHeaders)
+	if err != nil {
+		return "", fmt.Errorf("marshal Protected JWS Headers: %w", err)
+	}
+
+	b64Headers := base64.RawURLEncoding.EncodeToString(byteHeaders)
+
+	b64Payload := ""
+	if !detached {
+		b64Payload = base64.RawURLEncoding.EncodeToString(s.Payload)
+	}
+
+	b64Signature := base64.RawURLEncoding.EncodeToString(s.signature)
+
+	return fmt.Sprintf("%s.%s.%s",
+		b64Headers,
+		b64Payload,
+		b64Signature), nil
+}
+
+// Signature returns a copy of JWS signature.
+func (s *JSONWebSignature) Signature() []byte {
+	if s.signature == nil {
+		return nil
+	}
+
+	return append(s.signature[:0:0], s.signature...)
```

Comment:
please elaborate on this.

Reply:
Substituted with the usage of conventional `copy()`.